### ISSUE TITLE
Failure in CGAL_SetupBoost.cmake with boost 1.70.0

### ISSUE THREAD

**Issue by fxcoudert** [CGAL/cgal#3851]
## Issue Details

On macOS (all versions), we are seeing the following issue with since the release of Boost 1.70.0. Any `cmake` invocation using `find_package(CGAL)` fails with:

```
-- The C compiler identification is AppleClang 10.0.1.10010046
-- The CXX compiler identification is AppleClang 10.0.1.10010046
-- Check for working C compiler: /usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang
-- Check for working C compiler: /usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang++
-- Check for working CXX compiler: /usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- __cplusplus is 199711
-- Found Boost 1.70.0 at /usr/local/lib/cmake/Boost-1.70.0
--   Requested configuration: QUIET REQUIRED COMPONENTS thread;system
-- Found boost_headers 1.70.0 at /usr/local/lib/cmake/boost_headers-1.70.0
-- Found boost_thread 1.70.0 at /usr/local/lib/cmake/boost_thread-1.70.0
-- No suitable boost_thread variant has been identified!
--   libboost_thread-mt.dylib (shared, BUILD_SHARED_LIBS not ON, set Boost_USE_STATIC_LIBS=OFF to override)
--   libboost_thread-mt.a (static, Boost_USE_STATIC_LIBS=OFF)
CMake Error at /usr/local/lib/cmake/Boost-1.70.0/BoostConfig.cmake:95 (find_package):
  Found package configuration file:

    /usr/local/lib/cmake/boost_thread-1.70.0/boost_thread-config.cmake

  but it set boost_thread_FOUND to FALSE so package "boost_thread" is
  considered to be NOT FOUND.  Reason given by package:

  No suitable build variant has been found.

Call Stack (most recent call first):
  /usr/local/lib/cmake/Boost-1.70.0/BoostConfig.cmake:124 (boost_find_dependency)
  /usr/local/Cellar/cmake/3.14.2_1/share/cmake/Modules/FindBoost.cmake:266 (find_package)
  /usr/local/lib/cmake/CGAL/CGAL_SetupBoost.cmake:48 (find_package)
  /usr/local/lib/cmake/CGAL/CGALConfig.cmake:87 (include)
  /usr/local/lib/cmake/CGAL/CGALConfig.cmake:142 (check_cgal_component)
  CMakeLists.txt:59 (find_package)
```

This was seen as part of [Homebrew testing for Boost 1.70.0](https://github.com/Homebrew/homebrew-core/pull/38890). We are using CGAL 4.14 and cmake 3.14.2 and the system compilers. The issue does not occur with Boost 1.69.0.

**Comment by scpeters** [comment]
That error happens when trying to compile against the existing `cgal` binary that works with boost 1.69.0. As a different test, I tried recompiling `cgal` against boost 1.70.0 and I see a linking error:

~~~
[ 25%] Linking CXX shared library ../../lib/libCGAL.dylib
cd /tmp/cgal-20190415-79807-1d3uxpb/CGAL-4.14/src/CGAL && /usr/local/Cellar/cmake/3.14.2/bin/cmake -E cmake_link_script CMakeFiles/CGAL.dir/link.txt --verbose=1
/usr/local/Homebrew/Library/Homebrew/shims/mac/super/clang++ -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk -dynamiclib -Wl,-headerpad_max_install_names -compatibility_version 13.0.0 -current_version 13.0.3 -o ../../lib/libCGAL.13.0.3.dylib -install_name /usr/local/lib/libCGAL.13.dylib CMakeFiles/CGAL.dir/all_files.cpp.o /usr/local/lib/libmpfr.dylib /usr/local/lib/libgmp.dylib 
Undefined symbols for architecture x86_64:
  "boost::detail::get_tss_data(void const*)", referenced from:
      CGAL::get_default_random() in all_files.cpp.o
      boost::thread_specific_ptr<CGAL::Random>::reset(CGAL::Random*) in all_files.cpp.o
  "boost::detail::set_tss_data(void const*, void (*)(void (*)(void*), void*), void (*)(void*), void*, bool)", referenced from:
      boost::thread_specific_ptr<CGAL::Random>::~thread_specific_ptr() in all_files.cpp.o
      boost::thread_specific_ptr<CGAL::Random>::reset(CGAL::Random*) in all_files.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/libCGAL.13.0.3.dylib] Error 1
make[1]: *** [src/CGAL/CMakeFiles/CGAL.dir/all] Error 2
make: *** [all] Error 2
~~~

The key is that `/usr/local/lib/libboost_thread-mt.dylib` is not listed on the linker line. Adding that library fixes that specific linking command. I'm guessing the [CGAL_detect_if_Boost_Thread_is_required function](https://github.com/CGAL/cgal/blob/c68cf8fc4c850f8cd84c6900faa781286a7117ed/Installation/cmake/modules/CGAL_SetupBoost.cmake#L20-L44) is not properly detecting that the boost thread libs are needed. I'll try setting `CGAL_requires_Boost_Thread` to true to see if that fixes the homebrew build.

**Comment by scpeters** [comment]
Hm...it's printing `__cplusplus is 199711` during configuration, which makes me think it's not using c++11? Here's the full cmake output:

* https://gist.github.com/scpeters/6aa378b1987f73985a27ca6c27c0c469

**Comment by maxGimeno** [comment]
Indeed, without c++11 the dylib for boost thread is needed. I just read that in boost 1.70, the cmake variable Boost_LIBRARIES is not set anymore. If this is true, as our cmake scripts for setting up boost calls
>  target_link_libraries(${target} ${keyword} ${Boost_LIBRARIES})

it could explain why the dylib for boost_thread is not linked. But in the cmake output you provided, it justs says that FindBoost (the one form cmake, not ours) simply didn't find the dylib. That would indicate a failure in boost cmake system to retrieve a lib from the .cmake. 
My problem is I don't reproduce this bug, I don't really know how to find out where the bug really is...

**Comment by scpeters** [comment]
@maxGimeno  can you share the cmake command you use and the cmake output when you configure cgal against boost 1.70? That might help with debugging.

**Comment by fxcoudert** [comment]
@maxGimeno I think it's this bug: https://lists.boost.org/Archives/boost/2019/02/245016.php
and I can find reports saying that `FindBoost` has changed behaviour, and does not work anymore on systems with static and shared libraries.

I don't understand `cmake` well enough to propose a fix, sadly.

**Comment by maxGimeno** [comment]
Oh, good catch ! 
A typical cmake command for configuring cgal would be:
>cmake -DBoost_INCLUDE_DIR=path/to/boost/include/dir -DCMAKE_CXX_FLAGS="-std=c++y" path/to/cgal/dir/containing/CGALConfig.cmake

With that cxx flag it should make boost usable in header-only, which SHOULD make the dylib useless and fix the bug, as it is supposed to work with a c++11 compiler (except for gcc I think, but clang works fine). If the dylib is still required, you need to add `-DBoost_THREAD_LIBRARY_RELEASE=path/to/lib`, and probably the same for any required lib, unless setting `Boost_LIBRARY_DIR_RELEASE` to the lib dir is enough.
I'll have to look in the tweak file to understand how it works, but if it is only a version to add t the list it's great :) 
Too bad I don't have a working mac to test it though ...

**Comment by lrineau** [comment]
In my opinion, there is no bug in `CGAL_SetupBoost.cmake`: the fault is the one of Boost-1.70: they have strongly coupled `BUILD_SHARED_LIBS` and `Boost_USE_STATIC_LIBS`.

However, it is easy to work around that bug of Boost. The pull-request #3886 now modifies the default for `Boost_STATIC_LIBS`, so that it is the opposite of `BUILD_SHARED_LIBS`.

**Comment by fxcoudert** [comment]
I've compiled CGAL 4.14 with the following patch:

```
diff -pur a/cmake/modules/CGAL_SetupBoost.cmake b/cmake/modules/CGAL_SetupBoost.cmake
--- a/cmake/modules/CGAL_SetupBoost.cmake	2019-03-28 21:00:14.000000000 +0100
+++ b/cmake/modules/CGAL_SetupBoost.cmake	2019-05-05 18:12:28.000000000 +0200
@@ -50,7 +50,7 @@ else()
   find_package( Boost 1.48 REQUIRED )
 endif()
 
-if(Boost_FOUND)
+if(Boost_FOUND AND Boost_VERSION VERSION_LESS 1.70)
   if(DEFINED Boost_DIR AND NOT Boost_DIR)
     # Unset that cache variable that is set in the cache by FindBoost
     # (while it was searching for boost-cmake).
@@ -92,6 +92,13 @@ function(use_CGAL_Boost_support target)
   if(NOT CGAL_Boost_USE_STATIC_LIBS AND CGAL_AUTO_LINK_ENABLED)
     target_compile_definitions(${target} ${keyword} BOOST_ALL_DYN_LINK=1)
   endif()
-  target_include_directories(${target} SYSTEM ${keyword} ${Boost_INCLUDE_DIRS})
-  target_link_libraries(${target} ${keyword} ${Boost_LIBRARIES})
+  if(TARGET Boost::boost)
+    target_link_libraries(${target} ${keyword} Boost::boost)
+    if (CGAL_requires_Boost_libs)
+      target_link_libraries(${target} ${keyword} Boost::thread)
+    endif()
+  else()
+    target_include_directories(${target} SYSTEM ${keyword} ${Boost_INCLUDE_DIRS})
+    target_link_libraries(${target} ${keyword} ${Boost_LIBRARIES})
+  endif()
 endfunction()
diff -pur a/cmake/modules/CGAL_SetupCGAL_CoreDependencies.cmake b/cmake/modules/CGAL_SetupCGAL_CoreDependencies.cmake
--- a/cmake/modules/CGAL_SetupCGAL_CoreDependencies.cmake	2019-03-28 21:00:14.000000000 +0100
+++ b/cmake/modules/CGAL_SetupCGAL_CoreDependencies.cmake	2019-05-05 18:15:03.000000000 +0200
@@ -54,8 +54,8 @@ endif()
 #
 
 # See the release notes of CGAL-4.10: CGAL_Core now requires
-# Boost.Thread, with all compilers but MSVC.
-if (NOT MSVC)
+# Boost.Thread, with GNU G++.
+if (CMAKE_CXX_COMPILER_ID STREQUAL GNU)
   find_package( Boost 1.48 REQUIRED COMPONENTS thread system )
 endif()
 
@@ -71,8 +71,8 @@ function(CGAL_setup_CGAL_Core_dependenci
   target_link_libraries( CGAL_Core ${keyword} CGAL::CGAL )
 
   # See the release notes of CGAL-4.10: CGAL_Core now requires
-  # Boost.Thread, with all compilers but MSVC.
-  if (NOT MSVC)
+  # Boost.Thread, with GNU G++.
+  if (CMAKE_CXX_COMPILER_ID STREQUAL GNU)
     if(TARGET Boost::thread)
       target_link_libraries( CGAL_Core ${keyword} Boost::thread)
     else()
diff -pur a/cmake/modules/CGAL_TweakFindBoost.cmake b/cmake/modules/CGAL_TweakFindBoost.cmake
--- a/cmake/modules/CGAL_TweakFindBoost.cmake	2019-03-28 21:00:14.000000000 +0100
+++ b/cmake/modules/CGAL_TweakFindBoost.cmake	2019-05-05 18:15:26.000000000 +0200
@@ -29,16 +29,19 @@ if( NOT CGAL_TweakFindBoost )
     cmake_policy(SET CMP0077 NEW)
   endif()
   if(DEFINED CGAL_Boost_USE_STATIC_LIBS)
-    # If the option is loaded from CGALConfig.h, use its value as default
+    # If the option is loaded from CGALConfig.cmake, use its value as default
     # value.  But the user will still have the choice to change the
     # value. That means that we can build the CGAL libraries using static
     # or shared Boost libraries, and after build programs using CGAL with a
     # different setting for Boost libraries.
     set(CGAL_Boost_USE_STATIC_LIBS_DEFAULT ${CGAL_Boost_USE_STATIC_LIBS})
   else()
-    # Else the option is OFF by default. That means the use of shared Boost
-    # libraries is the default.
-    set(CGAL_Boost_USE_STATIC_LIBS_DEFAULT OFF)
+    # Else the option default is related to BUILD_SHARED_LIBS.
+    if(BUILD_SHARED_LIBS OR NOT DEFINED BUILD_SHARED_LIBS)
+      set(CGAL_Boost_USE_STATIC_LIBS_DEFAULT OFF)
+    else()
+      set(CGAL_Boost_USE_STATIC_LIBS_DEFAULT ON)
+    endif()
   endif()
 
   option(Boost_DEBUG "Activate the debug messages of the script FindBoost" OFF)
@@ -62,6 +65,26 @@ if( NOT CGAL_TweakFindBoost )
   endif()
 
   set(Boost_ADDITIONAL_VERSIONS
+    "1.89.1" "1.89.0" "1.89"
+    "1.88.1" "1.88.0" "1.88"
+    "1.87.1" "1.87.0" "1.87"
+    "1.86.1" "1.86.0" "1.86"
+    "1.85.1" "1.85.0" "1.85"
+    "1.84.1" "1.84.0" "1.84"
+    "1.83.1" "1.83.0" "1.83"
+    "1.82.1" "1.82.0" "1.82"
+    "1.81.1" "1.81.0" "1.81"
+    "1.80.1" "1.80.0" "1.80"
+    "1.79.1" "1.79.0" "1.79"
+    "1.78.1" "1.78.0" "1.78"
+    "1.77.1" "1.77.0" "1.77"
+    "1.76.1" "1.76.0" "1.76"
+    "1.75.1" "1.75.0" "1.75"
+    "1.74.1" "1.74.0" "1.74"
+    "1.73.1" "1.73.0" "1.73"
+    "1.72.1" "1.72.0" "1.72"
+    "1.71.1" "1.71.0" "1.71"
+    "1.70.1" "1.70.0" "1.70"
     "1.69.1" "1.69.0" "1.69"
     "1.68.1" "1.68.0" "1.68"
     "1.67.1" "1.67.0" "1.67"
diff -pur a/include/CGAL/CORE/MemoryPool.h b/include/CGAL/CORE/MemoryPool.h
--- a/include/CGAL/CORE/MemoryPool.h	2019-03-28 21:03:48.000000000 +0100
+++ b/include/CGAL/CORE/MemoryPool.h	2019-05-05 18:11:47.000000000 +0200
@@ -37,7 +37,8 @@
 
 #include <CGAL/config.h>
 #include <CGAL/tss.h>
-#if CGAL_STATIC_THREAD_LOCAL_USE_BOOST || (defined(CGAL_HAS_THREADS)&&__GNUC__)
+#include <boost/config.hpp>
+#if CGAL_STATIC_THREAD_LOCAL_USE_BOOST || (defined(CGAL_HAS_THREADS) && BOOST_GCC)
 // Force the use of Boost.Thread with g++ and C++11, because of the PR66944
 //   https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66944
 // See also CGAL PR #1888
@@ -91,7 +92,7 @@ public:
 
   // Access the corresponding static global allocator.
   static MemoryPool<T,nObjects>& global_allocator() {
-#if CGAL_STATIC_THREAD_LOCAL_USE_BOOST || (defined(CGAL_HAS_THREADS)&&__GNUC__)
+#if CGAL_STATIC_THREAD_LOCAL_USE_BOOST || (defined(CGAL_HAS_THREADS) && BOOST_GCC)
     if(memPool_ptr.get() == NULL) {memPool_ptr.reset(new Self());}
     Self& memPool =  * memPool_ptr.get();
 #endif
@@ -102,7 +103,7 @@ private:
    Thunk* head; // next available block in the pool
   std::vector<void*> blocks;
 
-#if CGAL_STATIC_THREAD_LOCAL_USE_BOOST || (defined(CGAL_HAS_THREADS)&&__GNUC__)
+#if CGAL_STATIC_THREAD_LOCAL_USE_BOOST || (defined(CGAL_HAS_THREADS) && BOOST_GCC)
   static boost::thread_specific_ptr<Self> memPool_ptr;
 #elif defined(CGAL_HAS_THREADS) // use the C++11 implementation
   static thread_local Self memPool;
@@ -111,7 +112,7 @@ private:
 #endif // not CGAL_HAS_THREADS
 };
 
-#if CGAL_STATIC_THREAD_LOCAL_USE_BOOST || (defined(CGAL_HAS_THREADS)&&__GNUC__)
+#if CGAL_STATIC_THREAD_LOCAL_USE_BOOST || (defined(CGAL_HAS_THREADS) && BOOST_GCC)
 template <class T, int nObjects >
 boost::thread_specific_ptr<MemoryPool<T, nObjects> >
 MemoryPool<T, nObjects>::memPool_ptr;
```

which is ported from https://github.com/CGAL/cgal/pull/3886 (which does not apply cleanly to CGAL 4.14). However, while CGAL builds fine, its cmake files still cannot be used by downstream project, which fail with:

```
-- Found Boost 1.70.0 at /usr/local/lib/cmake/Boost-1.70.0
--   Requested configuration: QUIET REQUIRED COMPONENTS thread;system
-- Found boost_headers 1.70.0 at /usr/local/lib/cmake/boost_headers-1.70.0
-- Found boost_thread 1.70.0 at /usr/local/lib/cmake/boost_thread-1.70.0
-- No suitable boost_thread variant has been identified!
--   libboost_thread-mt.dylib (shared, BUILD_SHARED_LIBS not ON, set Boost_USE_STATIC_LIBS=OFF to override)
--   libboost_thread-mt.a (static, Boost_USE_STATIC_LIBS=OFF)
CMake Error at /usr/local/lib/cmake/Boost-1.70.0/BoostConfig.cmake:95 (find_package):
  Found package configuration file:

    /usr/local/lib/cmake/boost_thread-1.70.0/boost_thread-config.cmake

  but it set boost_thread_FOUND to FALSE so package "boost_thread" is
  considered to be NOT FOUND.  Reason given by package:

  No suitable build variant has been found.

Call Stack (most recent call first):
  /usr/local/lib/cmake/Boost-1.70.0/BoostConfig.cmake:124 (boost_find_dependency)
  /usr/local/Cellar/cmake/3.14.3/share/cmake/Modules/FindBoost.cmake:266 (find_package)
  /usr/local/Cellar/cgal/4.14_1/lib/cmake/CGAL/CGAL_SetupBoost.cmake:48 (find_package)
  /usr/local/Cellar/cgal/4.14_1/lib/cmake/CGAL/CGALConfig.cmake:87 (include)
  /usr/local/Cellar/cgal/4.14_1/lib/cmake/CGAL/CGALConfig.cmake:142 (check_cgal_component)
  CMakeLists.txt:2 (find_package)
```

**Comment by maxGimeno** [comment]
That's another issue due to boost 1.70's changes to their cmake management. A fix  is currently being tested and should be integrated soon.

**Comment by lrineau** [comment]
The PR #3886 is still being tested and improved. The changes in Boost CMake support have a lot of implications for CGAL CMake scripts. I am working on it. Once the PR is ready and fully tested, it will be integrated in release branches for CGAL-4.13.2 and CGAL-4.14.1, and in CGAL `master` branch. I will then deal with potential merge conflicts.

**Comment by BeamupAI** [comment]
Same thing on Windows... any workaround?

**Comment by lrineau** [comment]
Fixed by PR #3851.
